python/sparkdl/graph_factory.py

```diff
@@ -0,0 +1,168 @@
+            gfn = builder.asGraphFunction([pred_input], [final_output])
+
+        return gfn
+
+    @classmethod
+    def build_flattener(cls):
+        with GraphBuilderSession() as builder:
+            mat_input = tf.placeholder(tf.float32, [None, None])
+            mat_output = tf.identity(tf.reshape(mat_input, shape=[-1]), name='output')
+            gfn = builder.asGraphFunction([mat_input], [mat_output])
+
+        return gfn
+    
+    @classmethod
+    def import_bare_keras(cls, model_or_file_path):
+        """ Build an UDF from a Keras model
+        """
+        if isinstance(model_or_file_path, KerasModel):
+            model = model_or_file_path
+            model_path = Path(mkdtemp(prefix='kera-')) / "model.h5"
+            # Save to tempdir and restore in a new session
+            model.save(str(model_path), overwrite=True)
+            is_temp_model = True
+        else:
+            model_path = model_or_file_path
+            is_temp_model = False
+
+        # Keras load function requires path string
+        if not isinstance(model_path, six.string_types):
+            model_path = str(model_path)
+
+        with GraphBuilderSession(wrap_keras=True) as builder:
+            K.set_session(builder.sess)
+            K.set_learning_phase(0) # Testing phase
+            model = load_model(model_path)
+            if is_temp_model:
+                shutil.rmtree(str(Path(model_path).parent), ignore_errors=True)
+
+            gfn = builder.asGraphFunction(model.inputs, model.outputs)
+
+        return gfn
+
+    @classmethod
+    def pipeline(cls, functions):
+        """ 
+        Takes multiple graph functions and merges them into a single graph function.
+        It is assumed that there is only one input and one output in the intermediary layers
+
+        :param functions: a list of tuples (scope name, GraphFunction object).
+        """
+        assert len(functions) >= 1, ("must provide at least one function", functions)
+        if 1 == len(functions):
+            return functions[0]
+        for (scope_in, gfn_in), (scope_out, gfn_out) in zip(functions[:-1], functions[1:]):
+            assert len(gfn_in.output_names) == len(gfn_out.input_names), \
+                "graph function link {} -> {} require compatible layers".format(scope_in, scope_out)
+            if len(gfn_out.input_names) != 1:
+                raise NotImplementedError("Only support single input/output for intermediary layers")
+
+        # Acquire initial placeholders' properties
+        with GraphBuilderSession() as builder:
+            _, first_gfn = functions[0]
+            feeds, _ = builder.import_graph_function(first_gfn, name='')
+            first_input_info = []
+            for tnsr in feeds:
+                name = builder.op_name(tnsr)
+                first_input_info.append((tnsr.dtype, tnsr.shape, name))
+
+        # Build a linear chain of all the provide functions
+        with GraphBuilderSession() as builder: 
```

Comment:
In this session, the new first graph function is reloaded with its own name scope (to avoid potential name conflict). In the previous session, we acquire the tensor info of the original inputs. In this session, the original inputs are reconstructed.